Remotion/Web/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProvider.cs

```diff
@@ -0,0 +1,251 @@
+              stringBuilder.AppendLine();
+            }
+          }
+        }
+      }
+    }
+
+    private enum FileChangeType
+    {
+      Changed,
+      Created,
+      Deleted,
+      Renamed
+    }
+
+    private readonly IResourcesPhysicalPathLocator _resourcesPhysicalPathLocator;
+    private readonly IResourceFileDetailsAppender _resourceFileDetailsAppender;
+    private readonly IReadOnlyList<ResourceType> _resourceTypes;
+
+    private readonly HashSet<string> _resourceTypeLookup;
+
+    private readonly object _lock = new();
+
+    private string? _physicalPath;
+    private FileSystemWatcher? _fileSystemWatcher;
+
+    // Stores the current cacheKey or null if no cache key has been generated
+    // Set to null without the lock, set to not-null always in the lock.
+    private string? _cacheKey;
+
+    public ResourcePathBuilderBasedStaticResourceCacheKeyProvider (IHttpContextProvider httpContextProvider, ResourceRoot resourceRoot)
+        : this(
+            new ResourcesPhysicalPathLocator(
+                ArgumentUtility.CheckNotNull("httpContextProvider", httpContextProvider),
+                ArgumentUtility.CheckNotNull("resourceRoot", resourceRoot)),
+            new ResourceFileDetailsAppender(),
+            ResourceType.CacheableResourceTypes)
+    {
+    }
+
+    protected ResourcePathBuilderBasedStaticResourceCacheKeyProvider (
+        IResourcesPhysicalPathLocator resourcesPhysicalPathLocator,
+        IResourceFileDetailsAppender resourceFileDetailsAppender,
+        IReadOnlyList<ResourceType> resourceTypes)
+    {
+      ArgumentUtility.CheckNotNull(nameof(resourcesPhysicalPathLocator), resourcesPhysicalPathLocator);
+      ArgumentUtility.CheckNotNull(nameof(resourceFileDetailsAppender), resourceFileDetailsAppender);
+      ArgumentUtility.CheckNotNull(nameof(resourceTypes), resourceTypes);
+
+      _resourcesPhysicalPathLocator = resourcesPhysicalPathLocator;
+      _resourceFileDetailsAppender = resourceFileDetailsAppender;
+      _resourceTypes = resourceTypes;
+
+      _resourceTypeLookup = resourceTypes.Select(e => e.Name).ToHashSet(StringComparer.OrdinalIgnoreCase);
+    }
+
+    public bool HasCacheKey => _cacheKey != null;
+
+    /// <inheritdoc />
+    public string GetStaticResourceCacheKey ()
+    {
+      var cacheKey = _cacheKey;
+      if (cacheKey != null)
+        return cacheKey;
+
+      lock (_lock)
+      {
+        // We ensure here that the field is only read once since it can be reset without a lock,
+        // which could cause a null value return here.
+        cacheKey = _cacheKey;
+        if (cacheKey != null)
+          return cacheKey;
+
+        if (_physicalPath == null)
+        {
+          var physicalPath = _resourcesPhysicalPathLocator.GetResourcesPhysicalPath();
+          physicalPath = Path.GetFullPath(physicalPath);
+          if (!physicalPath.EndsWith("" + Path.DirectorySeparatorChar))
```

Comment:
why do we need `"" + Path.Dired..` instead of simply testing for `Path.Direc...`

edit:
ah, because of .NET Framework. Okay, please add a conditional compile because we won't find this code again. manually  compare the last char.